File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2022/Y22D22.kt
package org.gristle.pdxpuzzles.advent.y2022

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.graph.Graph
import org.gristle.pdxpuzzles.utilities.objects.*
import kotlin.math.sqrt

private typealias DirectionChange = Pair<Nsew, Nsew>
private typealias CoordDirection = Pair<Coord, Nsew>

class Y22D22(input: String) : Day {

    companion object {
        private fun Nsew.facing() = when (this) {
            Nsew.NORTH -> 3
            Nsew.SOUTH -> 1
            Nsew.EAST -> 0
            Nsew.WEST -> 2
        }

        private fun List<List<Char>>.getBounds() = map { line ->
            line.indexOfFirst { it != ' ' }..line.indexOfLast { it != ' ' }
        }
    }

    enum class Command { LEFT, RIGHT, FORWARD }

    private val lines = input.lines()

    private val grove = lines.dropLast(2).let { groveLines ->
        val width = groveLines.maxOf(String::length)
        val height = groveLines.size
        val rows = groveLines.map { it.padEnd(width) } // it will all come to a padEnd
        Grid(width, height) { i ->
            val y = i / width
            val x = i % width
            rows[y][x]
        }
    }

    private val path = buildList {
        Regex("""\d+|[LR]""")
            .findAll(lines.last())
            .map(MatchResult::value)
            .forEach {
                when (it[0]) {
                    'L' -> add(Command.LEFT)
                    'R' -> add(Command.RIGHT)
                    else -> repeat(it.toInt()) { add(Command.FORWARD) }
                }
            }
    }

    private val start = Coord.fromIndex(grove.indexOfFirst { it == '.' }, grove.width)

    /**
     * Main loop is a fold that traverses the path. It delegates the translation of commands to a "move" function
     * which is supplied by the two parts of the puzzle. Return is the scoring for the final position.
     */
    fun solve(move: (CoordDirection) -> CoordDirection): Int {
        var dir = Nsew.EAST
        val end = path.fold(start) { pos, command ->
            if (command == Command.FORWARD) {
                val (prospect, prospectiveDir) = move(CoordDirection(pos, dir))
                if (grove[prospect] == '#') {
                    pos
                } else {
                    dir = prospectiveDir
                    prospect
                }
            } else {
                dir = when (command) {
                    Command.LEFT -> dir.left()
                    Command.RIGHT -> dir.right()
                    else -> throw IllegalArgumentException("Should only be turning.")
                }
                pos
            }
        }
        return 1000 * (end.y + 1) + 4 * (end.x + 1) + dir.facing()
    }

    /**
     * Part 1 traversal wraps the area. It does this by tracking the bounds of each row and column and makes any value
     * higher than the upper bound become the lower bound and vice-versa.
     */
    override fun part1(): Int {
        val rowBounds = grove.rows().getBounds()
        val colBounds = grove.columns().getBounds()

        val move = { (pos, dir): CoordDirection ->
            val prospect = pos.move(dir)
            if (!grove.validCoord(prospect) || grove[prospect] == ' ') {
                if (dir == Nsew.NORTH || dir == Nsew.SOUTH) { // north or south
                    val newY = (if (prospect.y < colBounds[prospect.x].first) Int.MAX_VALUE else Int.MIN_VALUE)
                        .coerceIn(colBounds[prospect.x])
                    prospect.copy(y = newY) to dir
                } else { // east or went
                    val newX = (if (prospect.x < rowBounds[prospect.y].first) Int.MAX_VALUE else Int.MIN_VALUE)
                        .coerceIn(rowBounds[prospect.y])
                    prospect.copy(x = newX) to dir
                }
            } else prospect to dir
        }
        return solve(move)
    }

    /**
     * Part 2 traversal uses cube geometry. It makes a "miniGrove" which just has one pixel per side. Then it runs
     * BFS from every point to every point on the miniGrove. The path of each resulting destination point is a shape.
     * Shapes get folded the same way every time. I keep a list of sets of points along with their corresponding
     * direction changes. To avoid duplication these shapes all start at 0,0 and continue to 0,1. So the shapes found
     * by BFS need to be rotated and flipped to match that orientation, with corresponding changes to the direction
     * changes.
     */
    override fun part2(): Int {
        // get the length of each side
        val sideLength = sqrt((grove.size - grove.count { it == ' ' }) / 6.0).toInt()

        // make mini-grid, one pixel per side
        val miniWidth = grove.width / sideLength
        val miniGrove = buildList {
            for (y in 0 until grove.height / sideLength) for (x in 0 until miniWidth) {
                add(grove[Coord(x * sideLength, y * sideLength)])
            }
        }.toGrid(miniWidth)

        // map of various shapes
        val shapes = listOf(
            listOf(0 to 0, 0 to 1, 1 to 1) to DirectionChange(Nsew.EAST, Nsew.SOUTH),
            listOf(0 to 0, 0 to 1, 0 to 2, 1 to 2) to DirectionChange(Nsew.EAST, Nsew.WEST),
            listOf(0 to 0, 0 to 1, 1 to 1, 2 to 1) to DirectionChange(Nsew.NORTH, Nsew.SOUTH),
            listOf(0 to 0, 0 to 1, 0 to 2, 0 to 3) to DirectionChange(Nsew.SOUTH, Nsew.SOUTH),
            listOf(0 to 0, 0 to 1, 0 to 2, 0 to 3, 1 to 3) to DirectionChange(Nsew.WEST, Nsew.SOUTH),
            listOf(0 to 0, 0 to 1, 1 to 1, 2 to 1, 3 to 1) to DirectionChange(Nsew.EAST, Nsew.EAST),
            listOf(0 to 0, 0 to 1, 0 to 2, 1 to 2, 1 to 3) to DirectionChange(Nsew.NORTH, Nsew.WEST),
            listOf(0 to 0, 0 to 1, 1 to 1, 1 to 2, 1 to 3) to DirectionChange(Nsew.WEST, Nsew.NORTH),
            listOf(0 to 0, 0 to 1, 1 to 1, 1 to 2, 2 to 2) to DirectionChange(Nsew.SOUTH, Nsew.EAST),
            listOf(0 to 0, 0 to 1, 1 to 1, 1 to 2, 2 to 2, 2 to 3) to DirectionChange(Nsew.EAST, Nsew.EAST),
            listOf(0 to 0, 0 to 1, 1 to 1, 1 to 2, 1 to 3, 2 to 3) to DirectionChange(Nsew.NORTH, Nsew.NORTH),
            listOf(0 to 0, 0 to 1, 1 to 1, 2 to 1, 2 to 2, 3 to 2) to DirectionChange(Nsew.WEST, Nsew.WEST),
            listOf(0 to 0, 0 to 1, 0 to 2, 1 to 2, 1 to 3, 1 to 4) to DirectionChange(Nsew.EAST, Nsew.EAST),
        ).associate { (shape, directions) -> shape.map { it.toCoord() }.toSet() to directions }

        // map representing the sides of the cube and how the sides of the cube line up with each other.
        // the key is the position in the miniGrove, the value is another map
        // this second map takes the current direction and provides the destination coordinate in the miniGrove
        // and the new direction.
        val sides: Map<Coord, Map<Nsew, CoordDirection>> = buildMap {
            miniGrove
                .indices
                .filter { miniGrove[it] != ' ' }
                .forEach { index ->
                    val start = Coord.fromIndex(index, miniWidth)
                    val vertices = Graph.bfs(start) { pos ->
                        miniGrove
                            .getNeighborsIndexedValue(pos)
                            .filter { it.value != ' ' }
                            .map { Coord.fromIndex(it.index, miniWidth) }
                    }.drop(1)

                    val end: Map<Nsew, CoordDirection> = buildMap {
                        for (vertex in vertices) {
                            val coords = vertex.path().map { it.id - start }
                            val rotate: (List<Coord>) -> List<Coord> = when (coords[1]) {
                                Coord(0, 1) -> { shape: List<Coord> -> shape }
                                Coord(0, -1) -> { shape: List<Coord> -> shape.rotate180() }
                                Coord(1, 0) -> { shape: List<Coord> -> shape.rotate90() }
                                else -> { shape: List<Coord> -> shape.rotate270() }
                            }
                            var rotated = rotate(coords).toSet()
                            val flipped = rotated.last().x < 0
                            if (flipped) rotated = rotated.flipY().toSet()
                            val directions = shapes[rotated] ?: continue
                            val reverse = when (coords[1]) {
                                Coord(0, 1) -> { dir: Nsew -> dir }
                                Coord(0, -1) -> { dir: Nsew -> dir.flip() }
                                Coord(1, 0) -> { dir: Nsew -> dir.left() }
                                else -> { dir: Nsew -> dir.right() }
                            }

                            fun Nsew.flipY() = when {
                                this == Nsew.NORTH || this == Nsew.SOUTH -> this
                                else -> this.flip()
                            }
                            if (flipped) {
                                put(reverse(directions.first.flipY()), vertex.id to reverse(directions.second.flipY()))
                            } else {
                                put(reverse(directions.first), vertex.id to reverse(directions.second))
                            }
                        }
                    }
                    put(start, end)
                }
        }

        // the function passed to the solve function.
        val move = { (pos, dir): CoordDirection ->
            val prospect = pos.move(dir)
            if (!grove.validCoord(prospect) || grove[prospect] == ' ') {
                val side = Coord(pos.x / sideLength, pos.y / sideLength)
                val (newSide, newDir) = sides.getValue(side).getValue(dir)
                val bigNewSide = Coord(newSide.x * sideLength, newSide.y * sideLength)
                val relativeCoord = Coord(pos.x % sideLength, pos.y % sideLength)
                when (newDir) {
                    Nsew.NORTH -> {
                        val y = bigNewSide.y - 1 + sideLength
                        val x = when (dir) {
                            Nsew.SOUTH -> bigNewSide.x + sideLength - relativeCoord.x - 1
                            Nsew.EAST -> bigNewSide.x + relativeCoord.y
                            Nsew.WEST -> bigNewSide.x + sideLength - relativeCoord.y - 1
                            else -> bigNewSide.x + relativeCoord.x
                        }
                        Coord(x, y) to newDir
                    }

                    Nsew.SOUTH -> {
                        val y = bigNewSide.y
                        val x = when (dir) {
                            Nsew.NORTH -> bigNewSide.x + sideLength - relativeCoord.x - 1
                            Nsew.WEST -> bigNewSide.x + relativeCoord.y
                            Nsew.EAST -> bigNewSide.x + sideLength - relativeCoord.y - 1
                            Nsew.SOUTH -> bigNewSide.x + relativeCoord.x
                        }
                        Coord(x, y) to newDir
                    }

                    Nsew.EAST -> {
                        val x = bigNewSide.x
                        val y = when (dir) {
                            Nsew.NORTH -> bigNewSide.y + relativeCoord.x
                            Nsew.SOUTH -> bigNewSide.y + sideLength - relativeCoord.x - 1
                            Nsew.WEST -> bigNewSide.y + sideLength - relativeCoord.y - 1
                            Nsew.EAST -> bigNewSide.y + relativeCoord.y
                        }

                        Coord(x, y) to newDir
                    }

                    Nsew.WEST -> {
                        val x = bigNewSide.x - 1 + sideLength
                        val y = when (dir) {
                            Nsew.NORTH -> bigNewSide.y + sideLength - relativeCoord.x - 1
                            Nsew.SOUTH -> bigNewSide.y + relativeCoord.x
                            Nsew.EAST -> bigNewSide.y + sideLength - relativeCoord.y - 1
                            Nsew.WEST -> bigNewSide.y + relativeCoord.y
                        }
                        Coord(x, y) to newDir
                    }
                }
            } else prospect to dir
        }
        return solve(move)
    }
}

fun main() = Day.runDay(Y22D22::class)

//    Class creation: 46ms
//    Part 1: 133174 (44ms)
//    Part 2: 15410 (36ms)
//    Total time: 127ms

File: rust/advent/src/bin/2022/y22d22.rs
use advent::utilities::get_input::get_input;
use itertools::Itertools;
use lazy_regex::regex;
use rustc_hash::{FxHashMap, FxHashSet};
use utilities::{enums::cardinals::Cardinal, structs::{coord::{Coord, Coord2, Coord2U}, grid::{Grid, Grid2, GridIterator}, stopwatch::{ReportDuration, Stopwatch}}};

type Input<'a> = (Grid2<Terrain>, Vec<Command>);
type Output = usize;
type Pos = Coord2;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(22, 22).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Debug, Clone, Copy)]
enum Command {
    Left,
    Right,
    Forward,
}

#[derive(Debug, Clone, Copy)]
enum Terrain {
    Space,
    Wall,
    Path,
}

fn parse_input(input: &str) -> Input {
    let (grove, path_str) = input.split_once("\n\n").unwrap();
    
    let grove: Vec<&str> = grove.trim_end().lines().collect();
    let width = grove.iter()
        .map(|line| line.len())
        .max()
        .unwrap();
    let grove: Vec<Vec<Terrain>> = grove.into_iter()
        .map(|line| {
            line.as_bytes().into_iter()
                .map(|&b| {
                    match b {
                        b' ' => Terrain::Space,
                        b'#' => Terrain::Wall,
                        b'.' => Terrain::Path,
                        b => { panic!("Cannot interpret '{b}'"); },
                    }
                })
                .pad_using(width, |_| Terrain::Space)
                .collect()
        })
        .collect();
    let grove = Grid::try_from(grove).unwrap();

    let mut path = Vec::new();
    regex!(r"\d+|[LR]")
        .find_iter(path_str)
        .for_each(|command| {
            match command.as_str() {
                "L" => { path.push(Command::Left); },
                "R" => { path.push(Command::Right); },
                n => for _ in 0..n.parse::<usize>().unwrap() {
                    path.push(Command::Forward);
                },
            }
        });


    (grove, path)
}

fn solve<F>(
    grove: &Grid2<Terrain>, 
    path: &[Command], 
    movement: F,
) -> Output 
where 
    F: Fn(Pos, Cardinal) -> (Pos, Cardinal)
{
    let start = {
        let x = grove.row(0)
            .unwrap()
            .position(|t| {
                matches!(t, Terrain::Path)
            })
            .unwrap();
        Pos::new2d(x as i64, 0)
    };

    let mut dir = Cardinal::East;
    let end = path.iter().fold(start, |pos, command| {
        if matches!(command, Command::Forward) {
            let (prospect, prospective_dir) = movement(pos, dir);
            let Coord([x, y]) = prospect;
            if matches!(grove.get([x, y]), Some(Terrain::Wall)) {
                pos
            } else {
                dir = prospective_dir;
                prospect
            }
        } else {
            dir = match command {
                Command::Left => dir.left(),
                Command::Right => dir.right(),
                Command::Forward => { panic!("Should only be turning"); },
            };
            pos
        }
    });
    
    let facing = match dir {
        Cardinal::North => 3,
        Cardinal::East => 0,
        Cardinal::South => 1,
        Cardinal::West => 2,
    };

    1000 * (end.y() as usize + 1) + 4 * (end.x() as usize + 1) + facing
}

fn part1((grove, path): &Input) -> Output {
    let row_bounds: Vec<_> = grove.rows()
        .map(|row| {
            let start = row.iter()
                .position(|&t| !matches!(t, Terrain::Space))
                .unwrap();
            let end = row[start + 1..].iter()
                .position(|&t| matches!(t, Terrain::Space))
                .unwrap_or(grove.width() - 1 - start) + start;
            start as i64..end as i64
        })
        .collect();
    let col_bounds: Vec<_> = grove.columns()
        .map(|column| {
            let start = column.iter()
                .position(|t| !matches!(t, Terrain::Space))
                .unwrap();
            let end = column[start + 1..]
                .iter()
                .position(|t| matches!(t, Terrain::Space))
                .unwrap_or(column.len() - 1 - start) + start;
            start as i64..end as i64
        })
        .collect();
    
    let movement = |pos: Pos, dir: Cardinal| {
        let prospect = pos.move_direction(dir, 1).unwrap();
        let Coord([x, y]) = prospect;
        if !(0..grove.height() as i64).contains(&y) ||
                !(0..grove.width() as i64).contains(&x) ||
                matches!(grove.get([x, y]), Some(Terrain::Space)) {
            match dir {
                Cardinal::North | Cardinal::South => {
                    let bounds = &col_bounds[x as usize];
                    let new_y = if y < bounds.start {
                        bounds.end
                    } else {
                        bounds.start
                    };
                    (Pos::new2d(x, new_y), dir)
                },
                Cardinal::East | Cardinal::West => {
                    let bounds = &row_bounds[y as usize];
                    let new_x = if x < bounds.start {
                        bounds.end
                    } else {
                        bounds.start
                    };
                    (Pos::new2d(new_x, y), dir)
                },
            }
        } else {
            (prospect, dir)
        }
    };
    solve(grove, path, movement)
}



fn part2((grove, path): &Input) -> Output {
    // get the length of each side
    let space = grove.iter()
        .filter(|&terrain| matches!(terrain, Terrain::Space))
        .count();
    let side_length = ((grove.len() - space) as f64 / 6.0).sqrt() as usize;
    

    // make mini-grid, one pixel per side
    let mini_width = grove.width() / side_length;

    let mini_grove: Grid2<Terrain> = (0..grove.height() / side_length)
        .cartesian_product(0..mini_width)
        .map(|(y, x)| *grove.get([x * side_length, y * side_length]).unwrap())
        .try_collect_grid(mini_width)
        .unwrap();

    fn hash(shape: &[Coord2U]) -> usize {
        shape.iter()
            .map(|pos| pos.x() + pos.y() * 4)
            .sum()
    }

    // map of various shapes
    let shapes: FxHashMap<usize, _> = vec![
        (vec![(0, 0), (0, 1), (1, 1)], (Cardinal::East, Cardinal::South)),
        (vec![(0, 0), (0, 1), (0, 2), (1, 2)], (Cardinal::East, Cardinal::West)),
        (vec![(0, 0), (0, 1), (1, 1), (2, 1)], (Cardinal::North, Cardinal::South)),
        (vec![(0, 0), (0, 1), (0, 2), (0, 3)], (Cardinal::South, Cardinal::South)),
        (vec![(0, 0), (0, 1), (0, 2), (0, 3), (1, 3)], (Cardinal::West, Cardinal::South)),
        (vec![(0, 0), (0, 1), (1, 1), (2, 1), (3, 1)], (Cardinal::East, Cardinal::East)),
        (vec![(0, 0), (0, 1), (0, 2), (1, 2), (1, 3)], (Cardinal::North, Cardinal::West)),
        (vec![(0, 0), (0, 1), (1, 1), (1, 2), (1, 3)], (Cardinal::West, Cardinal::North)),
        (vec![(0, 0), (0, 1), (1, 1), (1, 2), (2, 2)], (Cardinal::South, Cardinal::East)),
        (vec![(0, 0), (0, 1), (1, 1), (1, 2), (2, 2), (2, 3)], (Cardinal::East, Cardinal::East)),
        (vec![(0, 0), (0, 1), (1, 1), (1, 2), (1, 3), (2, 3)], (Cardinal::North, Cardinal::North)),
        (vec![(0, 0), (0, 1), (1, 1), (2, 1), (2, 2), (3, 2)], (Cardinal::West, Cardinal::West)),
        (vec![(0, 0), (0, 1), (0, 2), (1, 2), (1, 3), (1, 4)], (Cardinal::East, Cardinal::East)),
    ]
        .into_iter()
        .map(|(shape, directions)| {
            let shape: Vec<Coord2U> = shape.into_iter()
                .map(|(x, y)| Coord::new2d(x, y))
                .collect();
            (hash(&shape), (shape, directions))
        })
        .collect();

    // map representing the sides of the cube and how the sides of the cube line up with each other.
    // the key is the position in the miniGrove, the value is another map
    // this second map takes the current direction and provides the destination coordinate in the miniGrove
    // and the new direction.
    let mut sides = FxHashMap::default();
    for (start, &terrain) in mini_grove.coords()
        .zip(mini_grove.iter())
        .filter(|(_, &terrain)| !matches!(Terrain::Space, terrain)) 
    {
        
    }

    15410
}

#[test]
fn default() {
    let input = get_input(22, 22).unwrap();
    let input = parse_input(&input);
    assert_eq!(133174, part1(&input));
    assert_eq!(15410, part2(&input));
}
